inc/ctest_2_cppunittest.h

```diff
@@ -0,0 +1,92 @@
+#define ALLOCATE_TESTDATA_SECTION_ATTRIBUTE __declspec(allocate("testdata$_C_attribute"))
+#define ALLOCATE_TESTDATA_SECTION_DAN __declspec(allocate("testdata$_dan"))
+
+// unfortunately we have to create the sections before using them
+// this code is also taken from CppUnitTest.h
+#pragma section("testvers$", read, shared)
+#pragma section("testdata$_A_class", read, shared)
+#pragma section("testdata$_B_method", read, shared)
+#pragma section("testdata$_C_attribute", read, shared)
+#pragma section("testdata$_dan", read, shared)
+
+// The reader of the metadata really really wants to have in the helperMethodName the class to which the "methods" belong
+// Based on reverse engineering the code at https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fvset%2FAgile%2FCppUnit%2FDiscoverer%2FTestMetaDataReaderForLatestVersion.cs&_a=contents&version=GBmain
+
+// these macros are used by ctest to generate extra metadata needed by CppUnitTest
+#define IMPLEMENT_CPP_UNIT_FIXTURE(infoTag, cppunittest_func_name, ctest_func_name, method_type) \
+__declspec(dllexport) CTEST_2_CPPUNITTEST_MemberMethodInfo* __stdcall MU_C2(GetTestMethodInfo_, cppunittest_func_name)() \
+{ \
+    ALLOCATE_TESTDATA_SECTION_METHOD static const CTEST_2_CPPUNITTEST_MethodMetadata MU_C2(CppUnitTestMethodMetadata_, cppunittest_func_name) = \
+        { L"" infoTag, L"" MU_TOSTRING(cppunittest_func_name), (const unsigned char*)"DummyTestClass::" MU_TOSTRING(cppunittest_func_name), (const unsigned char*)(__FUNCDNAME__) }; \
+    static CTEST_2_CPPUNITTEST_MemberMethodInfo s_Info = { method_type, {NULL}, &MU_C2(CppUnitTestMethodMetadata_, cppunittest_func_name) }; \
+    s_Info.method.pVoidStaticMethod = (void*)ctest_func_name; \
+    return &s_Info; \
+}
+
+#define CTEST_CUSTOM_TEST_SUITE_INITIALIZE_CODE(func_name) \
+    IMPLEMENT_CPP_UNIT_FIXTURE("TestClassInitializeInfo", func_name, TestSuiteInitialize, TestClassSetup)
+
+#define CTEST_CUSTOM_TEST_SUITE_CLEANUP_CODE(func_name) \
+    IMPLEMENT_CPP_UNIT_FIXTURE("TestClassCleanupInfo", func_name, TestSuiteCleanup, TestClassCleanup)
+
+#define CTEST_CUSTOM_TEST_FUNCTION_INITIALIZE_CODE(func_name) \
+    IMPLEMENT_CPP_UNIT_FIXTURE("TestMethodInitializeInfo", func_name, TestFunctionInitialize, TestMethodSetup)
+
+#define CTEST_CUSTOM_TEST_FUNCTION_CLEANUP_CODE(func_name) \
+    IMPLEMENT_CPP_UNIT_FIXTURE("TestMethodCleanupInfo", func_name, TestFunctionCleanup, TestMethodCleanup)
+
+#define CTEST_CUSTOM_TEST_FUNCTION_CODE(func_name) \
+__declspec(dllexport) CTEST_2_CPPUNITTEST_MemberMethodInfo* __stdcall MU_C2(GetTestMethodInfo_, func_name)() \
+{ \
+    ALLOCATE_TESTDATA_SECTION_METHOD static const CTEST_2_CPPUNITTEST_MethodMetadata MU_C2(CppUnitTestMethodMetadata_, func_name) = \
+        { L"TestMethodInfo", L"" MU_TOSTRING(func_name), (const unsigned char*)("DummyTestClass::__GetTestMethodInfo_" MU_TOSTRING(func_name)), (const unsigned char*)(__FUNCDNAME__), L"" __FILE__, __LINE__ }; \
```

Comment:
>, L"" __FILE__, __LINE__  [](start = 181, length = 25)

Is this the only reason `IMPLEMENT_CPP_UNIT_FIXTURE` isn't just used here instead?

Reply:
There was an obscure reason why it failed when I try to do that, I will add a comment

---
In reply to: [608148856](https://github.com/Azure/c-testrunnerswitcher/pull/65#discussion_r608148856) [](ancestors = 608148856)